### ISSUE TITLE
Add Support for the new GGUF format which replaces GGML

### ISSUE THREAD

**Issue by apcameron** [oobabooga/text-generation-webui#3676]
Llama.cpp has dropped support for the GGML format and now only supports GGUF

**Comment by berkut1** [comment]
https://github.com/abetlen/llama-cpp-python/issues/628
We can only wait.

**Comment by Ph0rk0z** [comment]
You can probably use: https://github.com/abetlen/llama-cpp-python/pull/633

if you merge it yourselves I would back up old llama-cpp-python. llama.cpp does not care about breaking changes.

**Comment by sirus20x6** [comment]
it looks like this is now done

**Comment by Yzord** [comment]
Still can't download them. Hope it will be supported soon :)

**Comment by sammcj** [comment]
It looks like we're now waiting on https://github.com/jllllll/llama-cpp-python-cuBLAS-wheels to be updated, I've added a request issue - https://github.com/jllllll/llama-cpp-python-cuBLAS-wheels/issues/3

**Comment by jllllll** [comment]
Wheels will be uploaded soon.

Keep in mind that this version of llama-cpp-python does not support GGML models. Only GGUF models.
ctransformers has already been updated in the webui to support GGUF.

Personally, I would prefer to wait until more GGML models are converted to GGUF before updating llama-cpp-python.

If ooba wants, I can implement an older version of llama-cpp-python as a separate package in order to maintain GGML support, but that is a pretty messy solution even if it is temporary.

**Comment by sirus20x6** [comment]
> Personally, I would prefer to wait until more GGML models are converted to GGUF before updating llama-cpp-python.

isn't there a conversion script? if so why wait?

**Comment by jllllll** [comment]
The conversion script is not guaranteed to work and it's usage can be somewhat involved in order to perform a proper conversion.
My concern with this is that there are many people using the webui that do not have the technical ability to run the conversion script. Updating now will force them to use ctransformers to load their GGML models, which also means they will run slower.

**Comment by sirus20x6** [comment]
yeah, I just deleted all my GGML files because GGUF came out. Guess I'll stick with ctransformers for now, but I think people want the ease of not having to set any parameters that comes with GGUF

**Comment by Dampfinchen** [comment]
> If ooba wants, I can implement the previous version of llama-cpp-python as separate packages in order to maintain GGML support, but that is a pretty messy solution even if it is temporary.

Yes, I do think this would be the ideal approach. Otherwise, many people are going to complain why it suddenly stops working. If a GGML file is detected, it just uses the older commit while GGUF would use the new one.

**Comment by sirus20x6** [comment]
what about just have the script convert them for people so no duplicate llama-cpp-python is needed? maybe a check and a "do you want to convert this to gguf?" ui element

**Comment by jllllll** [comment]
> > If ooba wants, I can implement the previous version of llama-cpp-python as separate packages in order to maintain GGML support, but that is a pretty messy solution even if it is temporary.
> 
> Yes, I do think this would be the ideal approach. Otherwise, many people are going to complain why it suddenly stops working. If a GGML file is detected, it just uses the older commit while GGUF would use the new one.

I have written the code needed to support this here: https://github.com/jllllll/text-generation-webui/commit/4a999e3bcd6f44e78eda3d448f4e36884aaf8966
I will make a PR for it if ooba wants to use it.

**Comment by jllllll** [comment]
> what about just have the script convert them for people so no duplicate llama-cpp-python is needed? maybe a check and a "do you want to convert this to gguf?" ui element

The conversion script is not guaranteed to work with every model.
One such model was just discovered in the Discord server: `MythoMax-L2-13B`

**Comment by Ph0rk0z** [comment]
Hope it works for those 70B I d/l over the last week. I think this is the 3rd or 4th time they deprecate a format and it's always all or nothing.

**Comment by jllllll** [comment]
Fortunately, GGUF is designed to be expandable. So, this should be the last format deprecation.

**Comment by Patronics** [comment]
I was just starting to experiment with installing local LLMs (I've wanted to experiment with them for ages but been too busy), but seems I've picked a tumultuous time to start, so eagerly waiting for this migration to be complete so I can download and start playing with the models that won't be obsolete in a week!

**Comment by FartyPants** [comment]
Works, but currently GGUF speak like Yoda:
Greetings, I am here to provide. Is there you need help with?

**Comment by Ph0rk0z** [comment]
What is this I hear about GGUFv2 and header compatibility removed by October?

**Comment by berkut1** [comment]
@Ph0rk0z yes, GGUFv2 was released, which is BC :) They decided it was better, because GGUFv1 is not so much yet.
So again, need to wait llama-cpp-python

**Comment by Ph0rk0z** [comment]
Right.. but they deprecate the format again? Why not deprecate the format now or keep it? Waiting till October so people upload already obsolete models?

**Comment by jllllll** [comment]
The GGUFv2 implementation is still compatible with v1. I don't think it will cause any issues.
v2 is pretty much just for 64-bit models.

**Comment by berkut1** [comment]
@Ph0rk0z well. someone says, that backward compatibility seems to work fine https://github.com/ggerganov/llama.cpp/pull/2821

**Comment by Ph0rk0z** [comment]
Right.. but I read this in the code

```

// NOTE: temporary handling of GGUFv1 >> remove after Oct 2023
static bool gguf_fread_str_cur(FILE * file, struct gguf_str * p, size_t * offset) {
    p->n    = 0;
    p->data = NULL;

    bool ok = true;
```

**Comment by jllllll** [comment]
Why on earth would they do that?
The whole point of GGUF was to eliminate the need for format deprecation:
https://github.com/philpax/ggml/blob/gguf-spec/docs/gguf.md#specification

**Comment by berkut1** [comment]
@Ph0rk0z Yes, I noticed that too when I checked the code. My first thoughts were based because of their initial discussion.

@jllllll I hope that is the last BC, it looks like their just forgot about 64 bit.

**Comment by Ph0rk0z** [comment]
Yea.. so people will upload GGUFv1 models next 2 months for what? It has only been a couple of days. Why not convert them now when there are few. But instead, no mention of this and we have to go looking for it.

I have low bandwidth so can't just re download every 70b.

Had this situation before, when GPU offloading was first created. All the GGML I downloaded had to be requantized from scratch or re-downloaded. You couldn't even use a script to convert them.

And I can't not bother with GGUF because some good PRs got merged after.

**Comment by mechanicmuthu** [comment]
They are saying in this [discussion](https://github.com/ggerganov/llama.cpp/pull/2821#issuecomment-1694511206) that its a simple command line conversion between GGUF v1 to GGUF v2. The deprecation in this GGUF format is simply for simplicity rather than having a technical reason to stop supporting GGUF v1.

**Comment by Ph0rk0z** [comment]
Right! Very simple! _Little chance of failure._

>I don't need --allow-requantize or --leave-output-tensor, right?

>It probably will work for the non-k-quants types but pretty sure k-quants won't work. (There were also some changes to the decisions k-quants makes for LLaMA2 70B models so in that particular case it wouldn't pass through all the tensors even if the other issues were dealt with.)


'spose I'll see how it goes for the current quants and turning them into GGUF.. but will they be GGUFv1 or v2 when I use the script now?

**Comment by Dharmavineta** [comment]
After days of troubleshooting and literally having brain freeze over the "can't load the model" error, I came across this thread now!!(RIP me) I burned about 30 gb downloading all kinds of models and none of them are gguf(talk about being unlucky). Will atleast the gguf version stay for a while now? I don't want to end up with unsupported versions and then again break my head over the same issue in a week time.

**Comment by Ph0rk0z** [comment]
Convert the GGML to GGUF if you got them recently it will work.
